pkg/client/listener.go

```diff
@@ -196,26 +196,33 @@ func (l *WorkflowRunsListener) RemoveWorkflowRun(
 }
 
 func (l *WorkflowRunsListener) retrySend(workflowRunId string) error {
-	l.clientMu.RLock()
-	defer l.clientMu.RUnlock()
+	for i := 0; i < DefaultActionListenerRetryCount; i++ {
+		l.clientMu.RLock()
+		client := l.client
+		l.clientMu.RUnlock()
 
-	if l.client == nil {
-		return fmt.Errorf("client is not connected")
-	}
+		if client == nil {
+			return fmt.Errorf("client is not connected")
+		}
 
-	for i := 0; i < DefaultActionListenerRetryCount; i++ {
-		err := l.client.Send(&dispatchercontracts.SubscribeToWorkflowRunsRequest{
+		err := client.Send(&dispatchercontracts.SubscribeToWorkflowRunsRequest{
 			WorkflowRunId: workflowRunId,
 		})
 
 		if err == nil {
 			return nil
 		}
 
+		l.l.Warn().Err(err).Msgf("failed to send workflow run subscription, attempt %d/%d", i+1, DefaultActionListenerRetryCount)
+
+		if retryErr := l.retrySubscribe(context.Background()); retryErr != nil {
+			l.l.Error().Err(retryErr).Msg("failed to resubscribe after send failure")
+		}
+
 		time.Sleep(DefaultActionListenerRetryInterval)
 	}
 
-	return fmt.Errorf("could not send to the worker after %d retries", DefaultActionListenerRetryCount)
+	return fmt.Errorf("could not send to the worker after %d retries, please restart the worker and try again", DefaultActionListenerRetryCount)
```

Comment:
this is confusing since it might not be a worker on the subscription end. should we be casing on this on the client directly?